plugins/chain_plugin/trx_finality_status_processing.cpp

```diff
@@ -1,12 +1,265 @@
+      }
+   }
+
+   void trx_finality_status_processing_impl::handle_rollback() {
+      const auto& indx = _storage.index().get<by_block_num>();
+      std::deque<chain::transaction_id_type> trxs;
+      auto iter = indx.lower_bound(_head_block_id);
+      std::transform(iter, indx.end(), std::back_inserter(trxs), []( const finality_status_object& obj ) { return obj.trx_id; });
+      for (const auto& trx_id : trxs) {
+         auto trx_iter = _storage.find(trx_id);
+         _storage.modify( trx_iter, []( finality_status_object& obj ) {
+            obj.forked_out = true;
+         } );
+      }
+   }
+
+   bool trx_finality_status_processing_impl::status_expiry_of_trxs(const fc::time_point& now) {
+      const auto& indx = _storage.index().get<by_status_expiry>();
+      std::deque<chain::transaction_id_type> remove_trxs;
+      const fc::time_point success_expiry = now - _success_duration;
+
+      // find the successful (in any block) transactions that are past the successful expiry time
+      auto success_iter = indx.upper_bound(boost::make_tuple(false, fc::time_point{}));
+      const auto fail_end = success_iter;
+      const auto success_end = indx.upper_bound(boost::make_tuple(true, success_expiry));
+      auto get_id = []( const finality_status_object& obj ) { return obj.trx_id; };
+      std::transform(success_iter, success_end, std::back_inserter(remove_trxs), get_id);
+
+      // find the failure (not in a block) transactions that are past the failure expiry time
+      auto fail_iter = indx.begin();
+      std::transform(fail_iter, fail_end, std::back_inserter(remove_trxs), get_id);
+
+      for (const auto& trx_id : remove_trxs) {
+         _storage.erase(trx_id);
+      }
+      return !remove_trxs.empty();
+   }
+
+   void trx_finality_status_processing_impl::ensure_storage(const fc::time_point& now) {
+      const int64_t remaining_storage = _max_storage - _storage.memory_size();
+      if (remaining_storage > 0) {
+         return;
+      }
+
+      auto percentage = [](uint64_t mem) {
+         const uint64_t pcnt = 90;
+         return (mem * pcnt)/100;
+      };
+      // determine how much we need to free to get back to at least the desired percentage of the storage
+      int64_t storage_to_free = _storage.memory_size() - percentage(_max_storage) - remaining_storage;
+      const auto& block_indx = _storage.index().get<by_block_num>();
+      const auto& status_expiry_indx = _storage.index().get<by_status_expiry>();
+      std::deque<chain::transaction_id_type> remove_trxs;
+
+      auto reduce_storage = [&storage_to_free,&remove_trxs](const finality_status_object& obj) {
+         storage_to_free -= obj.memory_size();
+         remove_trxs.push_back(obj.trx_id);
+      };
+
+      auto block_upper_bound = chain::block_id_type{};
+      // the end of the oldest failure section
+      const auto oldest_failure_end = status_expiry_indx.upper_bound( std::make_tuple( false, fc::time_point::now() ) );
```

Comment:
`now` is provided to this method, but not used.

Reply:
removed